Sources/SwiftInspectorVisitors/FileVisitor.swift

```diff
@@ -0,0 +1,108 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import Foundation
+import SwiftSyntax
+
+public final class FileVisitor: SyntaxVisitor {
+
+  public init(fileURL: URL) {
+    fileInfo = FileInfo(url: fileURL)
+  }
+
+  public private(set) var fileInfo: FileInfo
+
+  public override func visit(_ node: ImportDeclSyntax) -> SyntaxVisitorContinueKind {
+    let importVisitor = ImportVisitor()
+    importVisitor.walk(node)
+
+    fileInfo.appendImports(importVisitor.imports)
+
+    // We don't need to visit children because our visitor just did that for us.
+    return .skipChildren
+  }
+  public override func visit(_ node: ProtocolDeclSyntax) -> SyntaxVisitorContinueKind {
+    let protocolVisitor = ProtocolVisitor()
+    protocolVisitor.walk(node)
+
+    if let protocolInfo = protocolVisitor.protocolInfo {
+      fileInfo.appendProtocol(protocolInfo)
+    }
+
+    // We don't need to visit children because our visitor just did that for us.
+    return .skipChildren
+  }
+
+  public override func visit(_ node: StructDeclSyntax) -> SyntaxVisitorContinueKind {
+    let structVisitor = StructVisitor()
+    structVisitor.walk(node)
+
+    fileInfo.appendStructs(structVisitor.structs)
+    // TODO: append classes
+    // TODO: append enums
+
+    // We don't need to visit children because our visitor just did that for us.
+    return .skipChildren
+  }
+
+  public override func visit(_ node: ClassDeclSyntax) -> SyntaxVisitorContinueKind {
+    // TODO: find an append classes
+    // TODO: append structs
+    // TODO: append enums
+
+    // We don't need to visit children because our visitor just did that for us.
+    return .skipChildren
+  }
+
+  public override func visit(_ node: EnumDeclSyntax) -> SyntaxVisitorContinueKind {
+    // TODO: find an append enums
+    // TODO: append structs
+    // TODO: append classes
+
+    // We don't need to visit children because our visitor just did that for us.
+    return .skipChildren
+  }
+
+  // MARK: Private
+
+  private var hasFinishedParsingProtocol = false
```

Comment:
I don't think we need this?

Reply:
07cb398cf8d5ba28a260fb856dab81c5a919a1a9

Reply:
good catch 😉